### PR TITLE
Np 45413 search by category

### PR DESCRIPTION
**Pull request by EyobSikt**
Adding search by category to NVI candidates.

### CODE REVIEW

File: index-handlers/src/main/java/no/sikt/nva/nvi/index/aws/CandidateQuery.java [hunk 1]
```diff
@@ -50,19 +52,21 @@ public class CandidateQuery {
...
     private final String year;
+    private final String category;
 
     public CandidateQuery(List<URI> affiliations,
                           boolean excludeSubUnits,
                           QueryFilterType filter,
                           String username,
                           String customer,
-                          String year) {
+                          String year, String category) {
```
Comment (sondrev) [inline index-handlers/src/main/java/no/sikt/nva/nvi/index/aws/CandidateQuery.java:62]: seperate line

File: index-handlers/src/main/java/no/sikt/nva/nvi/index/aws/CandidateQuery.java [hunk 2]
```diff
@@ -161,7 +165,11 @@ private static Query yearQuery(String year) {
 
     }
 
+    private static Query categoryQuery(String category) {
+        return termQuery(Optional.ofNullable(category).orElse(EMPTY_STRING),
+                         jsonPathOf(PUBLICATION_DETAILS, TYPE, KEYWORD));
```
Comment (sondrev) [inline index-handlers/src/main/java/no/sikt/nva/nvi/index/aws/CandidateQuery.java:170]: The Optional should be "outside" the termQuery() instead. No need to create a termQuery for ""

File: index-handlers/src/test/java/no/sikt/nva/nvi/index/OpenSearchClientTest.java [hunk 3]
```diff
@@ -313,6 +314,19 @@ void shouldReturnSearchResultsUsingFilterAndSearchTermCombined(Entry<String, Int
...
+
+        var searchParameters =
+            defaultSearchParameters().withCategory(CATEGORY).withAffiliations(List.of(NTNU_INSTITUTION_ID)).build();
+
+        var searchResponse =
+            openSearchClient.search(searchParameters);
+
+        assertThat(searchResponse.hits().hits(), hasSize(1));
+    }
+
```
Comment (sondrev) [inline index-handlers/src/test/java/no/sikt/nva/nvi/index/OpenSearchClientTest.java:329]: Add a inverse test as well. Show that some other document will be excluded when searching for category

**Review by sondrev — CHANGES_REQUESTED** [review]

**Comment by sondrev** [comment]
@EyobSikt There is some functionality to give back URL for pagination which needs to be modified. See the test 
SearchNviCandidatesHandlerTest.shouldReturnPaginatedSearchResultWithCorrectQueryParamsFilterAndQueryInIdIfGiven

File: index-handlers/src/main/java/no/sikt/nva/nvi/index/aws/CandidateQuery.java (hunk 1 above)
Comment (EyobSikt) [reply index-handlers/src/main/java/no/sikt/nva/nvi/index/aws/CandidateQuery.java]: Done.

File: index-handlers/src/main/java/no/sikt/nva/nvi/index/aws/CandidateQuery.java (hunk 2 above)
Comment (EyobSikt) [reply index-handlers/src/main/java/no/sikt/nva/nvi/index/aws/CandidateQuery.java]: Done.

File: index-handlers/src/test/java/no/sikt/nva/nvi/index/OpenSearchClientTest.java (hunk 3 above)
Comment (EyobSikt) [reply index-handlers/src/test/java/no/sikt/nva/nvi/index/OpenSearchClientTest.java:330]: Done.

File: index-handlers/src/main/java/no/sikt/nva/nvi/index/SearchNviCandidatesHandler.java [hunk 4]
```diff
@@ -88,11 +88,13 @@ private CandidateSearchParameters getCandidateSearchParameters(RequestInfo reque
                                .orElse(List.of(topLevelOrg));
         var username = requestInfo.getUserName();
         var year = extractQueryParamPublicationDateOrDefault(requestInfo);
+        var category = Optional.ofNullable(extractQueryParamCategoryOrDefault(requestInfo))
+            .orElse(null);
```
Comment (sondrev) [inline index-handlers/src/main/java/no/sikt/nva/nvi/index/SearchNviCandidatesHandler.java:92]: why is this ofNullable? extractQueryParamCategoryOrDefault returns either the category as a string, OR an empty string "".
It cant return null

File: index-handlers/src/main/java/no/sikt/nva/nvi/index/aws/CandidateQuery.java [hunk 5]
```diff
@@ -244,6 +253,11 @@ private Optional<Query> createYearQuery(String year) {
 
     }
 
+    private Optional<Query> createCategoryQuery(String category) {
+        return nonNull(category) ? Optional.of(categoryQuery(category)) : Optional.empty();
```
Comment (sondrev) [inline index-handlers/src/main/java/no/sikt/nva/nvi/index/aws/CandidateQuery.java:257]: category can not be null as you have defined it

**Pull request merged**